### PR TITLE
[Security\Core] Fix NoopAuthenticationManager::authenticate() return value

### LINKED ISSUE #36804: Authenticators and always_authenticate_before_granting causes null TypeError
**Symfony version(s) affected**: 5.1.0-BETA1

**Description**  

When the new 5.1 authenticator security is activated and you have `always_authenticate_before_granting: true` configured, a "is granted" call will lead to a null TypeError in `TraceableAccessDecisionManager`:

```
Argument 1 passed to Symfony\Component\Security\Core\Authorization\TraceableAccessDecisionManager::decide() must implement interface Symfony\Component\Security\Core\Authentication\Token\TokenInterface, null given, called in vendor/symfony/symfony/src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php on line 62
```

This is happening because `AuthorizationChecker` is re-authenticating the token

https://github.com/symfony/security-core/blob/eadb75975ee3099ffd1ac8773f088198dbf126bf/Authorization/AuthorizationChecker.php#L58-L60

with a `NoopAuthenticationManager`

https://github.com/symfony/symfony/blob/be3a9a93f0354ecc86c9ed157f132e799cc912f5/src/Symfony/Component/Security/Http/Authentication/NoopAuthenticationManager.php

returning nothing, therefore returning implict `null`. This is actually against the contract of `AuthenticationManagerInterface`, which states in its PhpDoc `@return TokenInterface An authenticated TokenInterface instance, never null`

**How to reproduce**  

```yaml
security:
    always_authenticate_before_granting: true
    enable_authenticator_manager: true
```

Have this in a Twig template that is rendered or do a `isGranted()` call from PHP.

```twig
{% if is_granted("ROLE_USER") %}
```

**Possible Solution**  

As discussed with @wouterj on Slack:

- Return `$token` in `NoopAuthenticationManager` (not sure what the security implications could be)
- Or forbid `always_authenticate_before_granting` in combination with the new authenticator security

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #36804 
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Comment by wouterj** [comment]
Imho, we should do both suggestions of #36804: The `always_authenticate_before_granting` setting is completely ignored in the new system. It's better to make this explicit by not allowing it to be true when the new system is used.

Rationale: This option is heavily misused to reload user roles on the fly, which is not the intention of this option. This use-case has to be addressed by a new feature in Symfony Security, fixing all related issues with changing roles on the fly (such as being logged out since 4.4). As this was considered out of scope for the authenticator system PR, it was decided that if one needs to reload on the fly, one needs to stick with the old system in 5.1.

**Comment by chalasr** [comment]
@wouterj Makes sense, exception added. Thanks.
Suggestions welcome to improve the error message

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 1]
```diff
@@ -111,6 +111,10 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('security_rememberme.xml');
 
         if ($this->authenticatorManagerEnabled = $config['enable_authenticator_manager']) {
+            if ($config['always_authenticate_before_granting']) {
+                throw new InvalidConfigurationException('The security options "always_authenticate_before_granting" cannot be used when "enable_authenticator_manager" is set to true.');
```
Comment (wouterj) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:115]: Minor plural->singular note and I think we should point to the solution: Not using the new system (for now). Now I've flipped the review and I'm open to any improvement on my suggestion 😄 

```suggestion
                throw new InvalidConfigurationException('The security option "always_authenticate_before_granting" cannot be used when "enable_authenticator_manager" is set to true. If you rely on this behavior, use "enable_authenticator_manager: false".');
```

**Review by wouterj — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @chalasr.

**Pull request merged**